graphix/parameter.py

```diff
@@ -0,0 +1,213 @@
+        else:
+            return NotImplemented
+
+    def __rsub__(self, other) -> ParameterExpression:
+        if isinstance(other, numbers.Number):
+            return ParameterExpression(other - self._expression)
+        elif isinstance(other, ParameterExpression):
+            return ParameterExpression(other._expression - self._expression)
+        else:
+            return NotImplemented
+
+    def __neg__(self) -> ParameterExpression:
+        return ParameterExpression(-self._expression)
+
+    def __truediv__(self, other) -> ParameterExpression:
+        if isinstance(other, numbers.Number):
+            return ParameterExpression(self._expression / other)
+        elif isinstance(other, ParameterExpression):
+            return ParameterExpression(self._expression / other._expression)
+        else:
+            return NotImplemented
+
+    def __rtruediv__(self, other) -> ParameterExpression:
+        if isinstance(other, numbers.Number):
+            return ParameterExpression(other / self._expression)
+        elif isinstance(other, ParameterExpression):
+            return ParameterExpression(other._expression / self._expression)
+        else:
+            return NotImplemented
+
+    def __mod__(self, other) -> float:
+        """mod magic function returns nan so that evaluation of
+        mod of measurement angles in :meth:`graphix.pattern.is_pauli_measurement`
+        will not cause error. returns nan so that this will not be considered Pauli measurement.
+        """
+        assert isinstance(other, float) or isinstance(other, int)
+        return np.nan
+
+    def sin(self) -> ParameterExpression:
+        return ParameterExpression(sp.sin(self._expression))
+
+    def cos(self) -> ParameterExpression:
+        return ParameterExpression(sp.cos(self._expression))
+
+    def tan(self) -> ParameterExpression:
+        return ParameterExpression(sp.tan(self._expression))
+
+    def arcsin(self) -> ParameterExpression:
+        return ParameterExpression(sp.asin(self._expression))
+
+    def arccos(self) -> ParameterExpression:
+        return ParameterExpression(sp.acos(self._expression))
+
+    def arctan(self) -> ParameterExpression:
+        return ParameterExpression(sp.atan(self._expression))
+
+    def exp(self) -> ParameterExpression:
+        return ParameterExpression(sp.exp(self._expression))
+
+    def log(self) -> ParameterExpression:
+        return ParameterExpression(sp.log(self._expression))
+
+    def conjugate(self) -> ParameterExpression:
+        return ParameterExpression(sp.conjugate(self._expression))
+
+    def sqrt(self) -> ParameterExpression:
+        return ParameterExpression(sp.sqrt(self._expression))
+
+    @property
+    def expression(self) -> sp.Expr:
+        return self._expression
+
+    def __repr__(self) -> str:
+        return str(self._expression)
+
+    def subs(self, variable, value) -> ParameterExpression | numbers.Number:
```

Comment:
Please use `complex` instead of abstract `numbers.Number` .

Reply:
Numbers from `numpy` (or `sympy`...) are not necessarily instances of `complex`, for instance:
```
isinstance(np.array([1])[0], complex) == False
isinstance(np.array([1])[0], numbers.Number) == True
isinstance(sp.parse_expr(1), complex) == False
isinstance(sp.parse_expr(1), numbers.Number) == True
```

Reply:
So do you mean this block can sometimes return non-`complex` value?

```python
if isinstance(result, numbers.Number):
    return complex(result)
```

Reply:
If you're sure that `complex(result)` is always `complex` type, please use `ParameterExpression | complex`.

I personally believe that return type annotation should be as detailed as possible, or we need extra type narrowings.

Reply:
Oops, sorry, I answered without thinking enough: you are right, and after the coercion, the result is `complex` indeed.